assets/js/blocks/cart-checkout/checkout/form/address-step.js

```diff
@@ -0,0 +1,96 @@
+/**
+ * External dependencies
+ */
+import { useMemo } from '@wordpress/element';
+import PropTypes from 'prop-types';
+import { useCheckoutAddress } from '@woocommerce/base-hooks';
+import { useShippingDataContext } from '@woocommerce/base-context';
+
+/**
+ * Internal dependencies
+ */
+import BillingFieldsStep from './billing-fields-step';
+import ContactFieldsStep from './contact-fields-step';
+import ShippingFieldsStep from './shipping-fields-step';
+import './style.scss';
+
+const AddressStep = ( {
+	requireCompanyField,
+	requirePhoneField,
+	showApartmentField,
+	showCompanyField,
+	showPhoneField,
+} ) => {
+	const {
+		defaultAddressFields,
+		billingFields,
+		setBillingFields,
+		setEmail,
+		setPhone,
+		setShippingAsBilling,
+		setShippingFields,
+		shippingAsBilling,
+		shippingFields,
+		showBillingFields,
+	} = useCheckoutAddress();
+	const { needsShipping } = useShippingDataContext();
+	const addressFieldsConfig = useMemo( () => {
+		return {
+			company: {
+				...defaultAddressFields.company,
+				hidden: ! showCompanyField,
+				required: requireCompanyField,
+			},
+			address_2: {
+				...defaultAddressFields.address_2,
+				hidden: ! showApartmentField,
+			},
+		};
+	}, [
+		defaultAddressFields,
+		showCompanyField,
+		requireCompanyField,
+		showApartmentField,
+	] );
+
+	return (
+		<>
+			<ContactFieldsStep
+				emailValue={ billingFields.email }
+				onChangeEmail={ setEmail }
+			/>
+			{ needsShipping && (
+				<ShippingFieldsStep
+					addressFieldsConfig={ addressFieldsConfig }
+					billingFields={ billingFields }
+					defaultAddressFields={ defaultAddressFields }
+					requirePhoneField={ requirePhoneField }
+					setPhone={ setPhone }
+					setShippingAsBilling={ setShippingAsBilling }
+					setShippingFields={ setShippingFields }
+					shippingAsBilling={ shippingAsBilling }
+					shippingFields={ shippingFields }
+					showPhoneField={ showPhoneField }
```

Comment:
I feel like we're passing too many props to each component, maybe we can move the hook to inside `ShippingFieldsStep` and `BillingFieldsStep`, this is unrelated to this PR, just an idea.

Reply:
I found some weird behavior when calling `useCheckoutAddress()` from `ShippingFieldsStep` and `BillingFieldsStep` separately, that's why I moved it one level up so the hook is shared by both components. I will proceed merging as-is, we can investigate it further in the future.